### PR TITLE
Local KNN Attention: parallel local pathway in TransolverBlock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Global slice attention in Transolver captures long-range dependencies effectively but treats all spatial scales uniformly — a node's interaction with its immediate neighbors is mediated through the same 96 global slices as its interaction with distant nodes. In CFD, boundary layer physics and wake interactions are dominated by *local* spatial gradients (pressure gradients across the airfoil surface, velocity gradients in the wake). Adding a local KNN attention pathway that operates on each node's k-nearest neighbors should capture these fine-scale interactions more faithfully, particularly improving p_tan where the aft foil sits in the fore foil's wake and local flow structure matters most.

This is architecturally orthogonal to all current improvements (loss formulation, augmentation, spatial bias routing). The local pathway runs in parallel with global slice attention and is gated so it can be learned to zero if unhelpful.

## Instructions

Make the following changes to `cfd_tandemfoil/train.py`:

### Step 1: Add command-line flags

Add these argparse arguments:
```python
parser.add_argument('--local_knn_attn', action='store_true', help='Add local KNN attention pathway parallel to slice attention')
parser.add_argument('--local_knn_k', type=int, default=16, help='Number of nearest neighbors for local attention')
```

### Step 2: Create the LocalKNNAttention module

Add this class before `TransolverBlock`:

```python
class LocalKNNAttention(nn.Module):
    """Local attention over k-nearest spatial neighbors."""
    def __init__(self, n_hidden, n_heads, k=16):
        super().__init__()
        self.k = k
        self.n_heads = n_heads
        self.head_dim = n_hidden // n_heads
        self.scale = self.head_dim ** -0.5
        self.qkv = nn.Linear(n_hidden, 3 * n_hidden)
        self.out_proj = nn.Linear(n_hidden, n_hidden)
        # Learnable gate initialized to zero — starts as identity
        self.gate = nn.Parameter(torch.zeros(1))
        
    def forward(self, x, coords):
        """
        x: (B, N, C) — hidden features
        coords: (B, N, 2) — spatial coordinates for KNN
        """
        B, N, C = x.shape
        k = min(self.k, N)
        
        # Compute pairwise distances and find k-nearest neighbors
        # coords: (B, N, 2)
        with torch.no_grad():
            dists = torch.cdist(coords, coords)  # (B, N, N)
            _, knn_idx = dists.topk(k, dim=-1, largest=False)  # (B, N, k)
        
        # Compute Q, K, V
        qkv = self.qkv(x).reshape(B, N, 3, self.n_heads, self.head_dim)
        q, k_proj, v = qkv.unbind(dim=2)  # each (B, N, H, D)
        
        # Gather K, V for neighbors
        knn_idx_expanded = knn_idx.unsqueeze(-1).unsqueeze(-1).expand(B, N, k, self.n_heads, self.head_dim)
        k_local = k_proj.unsqueeze(2).expand(B, N, N, self.n_heads, self.head_dim).gather(2, knn_idx_expanded)  # (B, N, k, H, D)
        v_local = v.unsqueeze(2).expand(B, N, N, self.n_heads, self.head_dim).gather(2, knn_idx_expanded)  # (B, N, k, H, D)
        
        # Local attention: Q(B,N,H,D) @ K_local(B,N,k,H,D) -> (B,N,H,k)
        q = q.unsqueeze(2)  # (B, N, 1, H, D)
        attn = (q * k_local).sum(-1) * self.scale  # (B, N, k, H)
        attn = attn.softmax(dim=2)  # softmax over k neighbors
        
        # Weighted sum of local values
        out = (attn.unsqueeze(-1) * v_local).sum(2)  # (B, N, H, D)
        out = out.reshape(B, N, C)
        out = self.out_proj(out)
        
        return out * self.gate.tanh()  # gated output, starts near zero
```

**IMPORTANT memory optimization**: The naive `expand` + `gather` above may OOM for large N. Use this chunked fallback if N > 4096:

```python
# Instead of expanding the full (B, N, N, H, D) tensor, gather directly:
# k_local = torch.gather(k_proj, 1, knn_idx_expanded_alt)  where knn_idx is reshaped appropriately
# Specifically:
knn_flat = knn_idx.reshape(B, N * k)  # (B, N*k)
knn_flat_exp = knn_flat.unsqueeze(-1).unsqueeze(-1).expand(B, N * k, self.n_heads, self.head_dim)
k_local = k_proj.gather(1, knn_flat_exp).reshape(B, N, k, self.n_heads, self.head_dim)
v_local = v.gather(1, knn_flat_exp).reshape(B, N, k, self.n_heads, self.head_dim)
```

**Use this gather-based approach instead of the expand approach** — it's O(N*k) memory, not O(N^2).

### Step 3: Integrate into TransolverBlock

In `TransolverBlock.__init__()`, add after the existing attention setup:
```python
if getattr(args, 'local_knn_attn', False):
    self.local_attn = LocalKNNAttention(n_hidden, n_heads, k=getattr(args, 'local_knn_k', 16))
else:
    self.local_attn = None
```

In `TransolverBlock.forward()`, modify the attention residual computation. Currently (non-AdaLN path, ~line 411):
```python
fx = _ln(self.ln_1_post, self.attn(_ln(self.ln_1, fx), spatial_bias=sb, tandem_mask=tandem_mask, zone_features=zone_features) + fx)
```

Change to:
```python
attn_out = self.attn(_ln(self.ln_1, fx), spatial_bias=sb, tandem_mask=tandem_mask, zone_features=zone_features)
if self.local_attn is not None:
    attn_out = attn_out + self.local_attn(_ln(self.ln_1, fx), raw_xy)
fx = _ln(self.ln_1_post, attn_out + fx)
```

Do the same for the AdaLN path if it exists.

### Step 4: Wire up in Transolver constructor

In the `Transolver` class constructor where blocks are created, ensure `args` is passed through so each `TransolverBlock` can access `local_knn_attn` and `local_knn_k`. The `raw_xy` (spatial coordinates) should be the first 2 columns of the input features (x[:, :, :2]).

### Step 5: Ensure coords are passed through forward

In `Transolver.forward()`, extract `coords = x[:, :, :2]` (the raw xy coordinates) and pass it through to each `TransolverBlock.forward()`. The blocks then pass `raw_xy=coords` to the local attention module.

### Step 6: Run experiments

Run 2 seeds with `--local_knn_attn --local_knn_k 16`:

**Seed 42:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/local-knn-attn-k16" \
  --wandb_group alphonse-local-knn-attn \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --local_knn_attn --local_knn_k 16 \
  --seed 42
```

**Seed 73:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/local-knn-attn-k16" \
  --wandb_group alphonse-local-knn-attn \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --local_knn_attn --local_knn_k 16 \
  --seed 73
```

### Notes
- The zero-initialized gate ensures the model starts from the exact same function as baseline — local attention contribution is learned from zero.
- Monitor VRAM usage — the KNN computation on coords is O(N^2) for distance computation. If it OOMs, try `--local_knn_k 8` as fallback.
- If K=16 works, consider a follow-up with K=32 or K=8 for ablation.
- `raw_xy` should be `x[:, :, :2]` which are the (x, y) spatial coordinates already present in the input features.

## Baseline (current, PR #2184 — DCT Freq Loss w=0.05)

| Metric | 2-seed avg | Merge threshold |
|--------|-----------|-----------------|
| p_in   | 13.205    | < 13.21         |
| p_oodc | 7.816     | < 7.82          |
| p_tan  | 28.502    | < 28.50         |
| p_re   | 6.453     | < 6.45          |

W&B baseline runs: `6yfv5lio` (seed 42), `etepxvjc` (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Local KNN Attention

### Implementation Note
The original PR design used true KNN (pairwise distance matrix) which is infeasible for this mesh size (N≈120k nodes → O(N²) = 14.4B distances per sample). After iterating through several approaches (chunked KNN, random candidate KNN, random sparse attention), the final implementation uses **strided anchor attention**:

1. Select M=512 evenly-spaced anchor nodes from the N=120k mesh
2. Run full self-attention among M anchors (M²=262k attention weights — fast)
3. Propagate updates to all N nodes via distance-weighted interpolation from k=16 nearest anchors

Gate parameter initialized to 0 so model starts identical to baseline.

### Results

| Config | p_in | p_oodc | p_tan | p_re | W&B Run | Epochs |
|--------|------|--------|-------|------|---------|--------|
| **Baseline (2-seed avg)** | **13.21** | **7.82** | **28.50** | **6.45** | 6yfv5lio, etepxvjc | ~155 |
| k=16, seed 42 | 20.0 | 11.2 | 32.0 | 8.5 | `bpgzy063` | 95 |
| k=16, seed 73 | 17.9 | 10.9 | 32.9 | 8.2 | `9x2f1pis` | 96 |

### Analysis

**Strong negative result.** All metrics are significantly worse than baseline (35-50% degradation on p_in, 12-15% on p_tan).

**Root causes:**

1. **Throughput degradation**: Only 95-96 epochs reached vs ~155 baseline (~38% fewer epochs). The anchor-to-all scatter step computes a (B, N, M) = (4, 120k, 512) distance matrix (~1GB) at every forward pass, 3× per forward (one per TransolverBlock). This overhead reduces total training signal.

2. **Strided anchors are not spatially adaptive**: Evenly-spaced anchors don't concentrate where the mesh is dense (airfoil surface, wake). The hypothesis was about capturing *local* flow gradients at surfaces — but uniformly-spaced anchors mostly sample from the volume, missing the critical surface regions.

3. **Scatter interpolation is too coarse**: Propagating M=512 anchor outputs to N=120k nodes via nearest-anchor weighting means each anchor serves ~234 nodes on average. This is too coarse for capturing fine-scale boundary layer physics.

4. **Fundamental architecture mismatch**: The Transolver's slice attention already provides a learned spatial grouping mechanism via the physics slices. Adding a second attention pathway that uses a different (cruder) spatial decomposition may be counterproductive — it competes with rather than complements the slice routing.

### Suggested Follow-ups
- **Close this direction.** Local attention for N=120k meshes requires fundamentally different approaches than KNN/anchor sampling.
- If local interaction modeling is still desired, consider:
  - **Graph message passing** on a pre-computed KNN graph (compute once, reuse) rather than per-forward-pass
  - **Surface-only local attention** restricted to ~1-3k surface nodes (tractable for true KNN)
  - **Multi-scale slice counts** (e.g., 96 global + 256 fine-grained slices) to let the existing slice mechanism handle both scales

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2200 because ADVISOR: **Closing — clear dead end.**

All metrics 12-50% worse than baseline:
| Metric | Baseline | Seed 42 | Seed 73 | Δ avg |
|--------|----------|---------|---------|-------|
| p_in | 13.21 | 20.0 | 17.9 | +43% |
| p_oodc | 7.82 | 11.2 | 10.9 | +41% |
| p_tan | 28.50 | 32.0 | 32.9 | +14% |
| p_re | 6.45 | 8.5 | 8.2 | +30% |

Only 95-96 epochs reached (vs ~155 baseline) — the anchor-to-all scatter step (`(B, N, M) = (4, 120k, 512)`) adds massive overhead.

Good root cause analysis. You correctly identified that:
1. Strided anchors aren't spatially adaptive — they oversample volume, undersample critical surface regions
2. The 120k-node mesh makes true KNN attention computationally infeasible (O(N²))
3. The slice mechanism already provides learned spatial grouping — a second crude decomposition competes rather than complements

Your follow-up suggestion about **surface-only local attention** (restricted to ~1-3k surface nodes) is the most interesting direction from this experiment — noted for future consideration.

New assignment incoming..
